---
title: "BIOS 7659 Homework 8"
author: "Tim Vigers"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: word_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE,dpi=600,message=FALSE,warning=FALSE)
knitr::opts_knit$set(root.dir = "/Users/timvigers/GitHub/School/Statistical Genomics/hw8")
library(FlowSorted.Blood.450k)
library(quadprog)
library(IlluminaHumanMethylation450kmanifest)
library(bPeaks)
library(limma)
library(flextable)
library(tidyverse)
library(parallel)
set.seed(1017)
```

# 1. Cell-Type Composition

Read in the data:

```{r}
baseDir1 = "blood/plate1"
targets1 = read.metharray.sheet(baseDir1)
baseDir2 = "blood/plate2"
targets2 = read.metharray.sheet(baseDir2)
targets = rbind(targets1, targets2)
rgSet = read.metharray.exp(targets=targets,extended=T)
sampleNames(rgSet) = rgSet[[1]]
getManifest(rgSet)
clindat = read.table("blood/demographic.txt", sep="\t", header=T)
pData(rgSet)$Sample_Group = clindat$Exposure
pData(rgSet)$child_sex = clindat$child_sex
```

## a) Find differentially methylated positions based on exposure status

```{r}
# SWAN normalization
msetSWAN = preprocessSWAN(rgSet)
dmp = dmpFinder(msetSWAN,pheno = msetSWAN$Sample_Group,type = "continuous")
```

Here the unexposed group is the reference, so a negative value of beta in the DMP table indicates that the probe was hypomethylated in the exposed group (and a positive value indicates hypermethylation). There are no probes significant at the $q<0.1$ level, but there are `r sum(dmp$pval<1e-5,na.rm = T)` significant at the $p<10^{-5}$ level:

```{r}
dmp %>% rownames_to_column(var="Probe") %>% filter(pval < 1e-5) %>% mutate(Direction = ifelse(sign(beta)==-1,"Hypomethylated","Hypermethylated")) %>% select(Probe,beta,Direction,pval) %>% rename("Beta"=beta,"P Value"=pval) %>% flextable(.) %>% footnote(.,i=1,j=3,value = as_paragraph("With respect to exposure"),ref_symbols = "*",part = "header") %>% set_table_properties(.,layout = "autofit",width = 0.75)
```

## b) Cell-type composition

Cell-type composition is important in methylation studies because methylation is a key factor in converting cells from a pluripotent progenitor to specific cell-type. As a result, differentially methylated regions (DMRs) can distinguish cell-types with high sensitivity and specificity [1]. In other words, if you are comparing two samples with different cell-type composition without some sort of adjustment, many DMRs that distinguish between them will likely be the result of cell-type composition rather than the exposure of interest. There are many ways to adjust for cell-type composition when cell-type isn't directly measured, but the approach in Houseman et al. [1] is implemented in `estimateCellCounts()` in the `minfi` package, and that paper is frequently cited. The approach in Jaffe & Irizarry [2] is simply an adaptation of the Houseman algorithm for the 450K array. These approaches require reference data, but there are also reference-free methods such as singular value decomposition, linear models with principal components, SVA, RUV, RUVm, and others. However, the best option if possible is to measure complete blood counts rather than estimating cell types. 

The Houseman algorithm takes an $m\times 1$ vector of methylation values $\mathbf{Y}_{0h}$ for a homogenous cell population, and proposes the linear model $\mathbf{Y}_{0h}=\mathbf{B}_0\mathbf{w}_{0h}+\mathbf{e}_{0h}$ where $\mathbf{w}_{h}$ is a $d_0\times 1$ covariate vector indicating cell-type and $\mathbf{B}_0$ is an $m\times d_0$ matrix. Also, $h$ indexes the specimen and $\mathbf{e}_{0h}$ is an error vector. Next, they propose a similar model for the data without cell-type measurements: $\mathbf{Y}_{1i}=\mathbf{B}_1\mathbf{z}_{i}+\mathbf{e}_{1i}$ with the linking regression model $\mathbf{B}_1=\mathbf{1}_{m\gamma_0^T}+\mathbf{B}_0\Gamma+\mathbf{U}$ where $i$ indexes samples of interest and $\mathbf{U}$ is a matrix of errors. The math gets fairly complicated after this, so I won't go into any more detail, but the points is that the cell-type mixture coefficients can be "recovered" from $\Gamma$ by estimating $\hat{\mathbf{B}_0}$ and $\hat{\mathbf{B}_1}$ via linear models using ordinary least squares, mixed effects models, limma, or surrogate variable analysis. The authors also recommend bootstrap procedures for estimating standard errors.

## c) Estimate cell counts

```{r fig.width=8,fig.height=8}
rgSet = read.metharray.exp(targets=targets,extended=F)
cells = estimateCellCounts(rgSet,meanPlot = T,compositeCellType = "Blood",cellTypes = c("CD8T","CD4T","NK","Bcell","Mono","Gran"))
cells %>% as.data.frame(.) %>% flextable(.) %>% set_table_properties(.,layout = "autofit")
```

The plot above shows the average methylation across cell-type differentiating probes, for our data and the reference data set ("FlowSorted.Blood.450k"). The means of the two datasets should be in a similar range. If not, this suggests large batch effects in the data and the composition estimates should not be trusted (the function automatically performs quantile normalization between the two datasets to reduce potential batch effects). This plot looks reasonable, so we can probably trust these cell-type estimates and repeat part a) with adjustment for cell-type composition. Because the cell proportions necessarily sum to 1, we technically should not be treating them as independent variables. One option to get around this issue is to simply drop one of the cell types from the model, so for the purposes of this homework I've decided not to include the NK cells because these generally had very small estimated proportions.

```{r}
# Format data
m = getM(msetSWAN)
cells = cbind(msetSWAN$Sample_Group,cells)
colnames(cells)[1] = "Sample_Group"
cells = cells[,-which(colnames(cells)=="NK")]
# Fit models with limma
d = model.matrix(~ Sample_Group + CD8T + CD4T + Bcell + Mono + Gran,as.data.frame(cells))
fit = lmFit(m,d)
sd = fit$stdev.unscaled
beta = fit$coefficients[,"Sample_Group"]
# Calculate p values
pval = 2*pt(abs(beta / (sd[,"Sample_Group"] * fit$sigma)),
         df = fit$df.residual,lower.tail = F)
qval = p.adjust(pval,"fdr")
dmp = cbind(fit$coefficients,pval)
dmp = cbind(dmp,qval)
```

One thing I noticed when fitting these models is that many of the regression estimates for cell type are really large with correspondingly large standard errors. When I removed them from the model and re-ran the code I got the same results as `dmpFinder()` in part a), so I don't think it's a programming issue. These estimates may not be a problem, but with more time I would probably investigate them a little. However, for this homework I am assuming that the estimates are okay given that we aren't interested in inference for cell type. Also, I did copy the idea of using `limma::lmFit()` from Karen Kanaster because it is so much faster than `lm()`, but the code above is my own.

There are `r sum(qval<0.1)` probes significant at the $q<0.1$ level, and `r sum(dmp[,"pval"]<1e-5)` probes significant at the $p<10^{-5}$ level:

```{r}
dmp %>% as.data.frame(.) %>% rownames_to_column(var="Probe") %>% filter(pval < 1e-5) %>% mutate(Direction = ifelse(sign(Sample_Group)==-1,"Hypomethylated","Hypermethylated")) %>% select(Probe,Sample_Group,Direction,pval) %>% rename("Beta"=Sample_Group,"P Value"=pval) %>% flextable(.) %>% footnote(.,i=1,j=3,value = as_paragraph("With respect to exposure"),ref_symbols = "*",part = "header") %>% set_table_properties(.,layout = "autofit",width = 0.75)
```

# INTERPRET RESULTS HERE

# 2. ChIP-Seq

Read in the data:

```{r}
data(yeastCDS)
allData = dataReading("./chipseq/tup1_IP.txt", "./chipseq/mock_IP.txt",yeastSpecies = yeastCDS$Saccharomyces.cerevisiae)
```

## a) What methods were used?

## b) Average number of sequences mapped

```{r}
baseLineIP = baseLineCalc(allData$IPdata[,ncol(allData$IPdata)])
baseLineControl = baseLineCalc(allData$controlData[,ncol(allData$controlData)])
```

The average number of sequences mapped in the Tup1 ChIP sample is `r baseLineIP` and the average number in the mock sample is `r baseLineControl`, so the Tup1 ChIP sample appears to have a lot more reads.

## c) Peak detection

```{r}
# Subset the data
ip = allData$IPdata[allData$IPdata$V1 == "chrV",]
control = allData$controlData[allData$controlData$V1 == "chrV",]
peaks = peakDetection(ip$V3,control$V3,"chrV",baseLineIP = baseLineIP,baseLineControl = baseLineControl,outputName = "bPeaks_results_mock")
```

The `peakDetection()` function found `r nrow(peaks)` peaks. 

## d)

## e) Peak location

```{r}
# Find peaks
peak_location = peakLocation("./bPeaks_results_mock.bed",yeastCDS$Saccharomyces.cerevisiae,outputName = "bPeaksLocation_mock")
# Read results back in
promoters = read.delim("./bPeaksLocation_mock_peakLocation_inPromoters.txt",header = F)
cds = read.delim("./bPeaksLocation_mock_peakLocation_inCDS.txt",header = F)
```

Out of the `r peak_location$numPeaks` peaks detected, `r length(unique(promoters$V1))` are in promoters and `r length(unique(cds$V1))` are in CDS (genes). 

## f) Repeat using the input IP sample

```{r}
# Read in data
allData = dataReading("./chipseq/tup1_IP.txt","./chipseq/input_IP.txt",yeastSpecies = yeastCDS$Saccharomyces.cerevisiae)
# Baseline
baseLineControl = baseLineCalc(allData$controlData[,ncol(allData$controlData)])
# Peak detection
control = allData$controlData[allData$controlData$V1 == "chrV",]
peaks = peakDetection(ip$V3,control$V3,"chrV",baseLineIP = baseLineIP,baseLineControl = baseLineControl)
# Peak locations
peak_location = peakLocation("./bPeaks_results.bed",yeastCDS$Saccharomyces.cerevisiae,outputName = "bPeaksLocation_input")
promoters = read.delim("./bPeaksLocation_input_peakLocation_inPromoters.txt",header = F)
cds = read.delim("./bPeaksLocation_input_peakLocation_inCDS.txt",header = F)
```

The average number of sequences mapped in the input IP sample is `r baseLineControl`, which is much closer the Tup 1 sample (`r baseLineIP`) than the mock sample was. With the new control data, the `peakDetection()` function found `r nrow(peaks)` peaks. Of these peaks, `r length(unique(promoters$V1))` are in promoters and `r length(unique(cds$V1))` are in genes, so at least 1 peak must be in both a gene and promoter.

# References

1. Houseman EA, Accomando WP, Koestler DC, et al. DNA methylation arrays as surrogate measures of cell mixture distribution. BMC Bioinformatics. 2012;13(1):86. doi:10.1186/1471-2105-13-86

2. Jaffe AE, Irizarry RA. Accounting for cellular heterogeneity is critical in epigenome-wide association studies. Genome Biol. 2014;15(2):R31. doi:10.1186/gb-2014-15-2-r31
