---
title: "BIOS 7659 Homework 7"
author: "Tim Vigers"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: word_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo=TRUE,dpi = 600)
library(shinyMethyl)
library(minfi) 
library(bumphunter) 
library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
library(arsenal)
library(flextable)
library(tidyverse)
```

# 1. DNA Methylation QC and Normalization (Illumina 450K)

Load the data:

```{r warning=FALSE,message=FALSE}
baseDir = "C:/Users/tim/Dropbox/Documents/School/Statistical Genomics/Homework Files/HW7/idats"
targets = read.metharray.sheet(baseDir)
rgSet = read.metharray.exp(targets = targets)
annotation(rgSet)
```

## a) Table 1

```{r results='asis'}
df = as.data.frame(pData(rgSet))
df = df[df$Status == "cancer",]
t1 = tableby(~ patient.age_at_initial_pathologic_diagnosis +
               patient.height + patient.weight + Sex +  patient.race,
             data = df)
summary(t1,labelTranslations = 
          list(patient.age_at_initial_pathologic_diagnosis = 
                 "Age at Diagnosis",
               patient.height = "Height",patient.weight = "Weight",
               patient.race = "Race"))
```

There are three unique subjects in this dataset, each with two samples (one primary tumor sample and one from normal solid tissue).

## b) Type I and II probes

```{r}
getManifest(rgSet)
```

There are 135,476 type I probes and 350,036 type II probes. Type I probes have two different sequences per CpG site, one for methylated and one for unmethylated CpGs. Type II probes use a two-color channel, which allows each probe to measure both methylated and unmethylated CpGs. As a result, type II probes take up half the physical space of type I probes. However, they have a lower dynamic range than type I probes, and are also more biased and less reproducible. 

## c) QC Plots

### By ID

```{r}
id = pData(rgSet)$id
densityPlot(rgSet,sampGroups = id)
densityBeanPlot(rgSet,sampGroups = id)
```

Unfortunately I get the error "'sampNames' is not a graphical parameter" when trying to use `sampNames` as requested in the prompt, so I used sample ID as the group ID. On first inspection, nothing in the density plot particularly jumped out at me, although the peak for sample 5 is a little higher than the other at the 1.0 end of the x axis. This sample also looks different from the others in the bean plot because its center is skewed further to the right than the others.

### By sample type

```{r}
stype = pData(rgSet)$sample_type
densityPlot(rgSet,sampGroups = stype,legend = F)
densityBeanPlot(rgSet,sampGroups = stype)
```

I removed the legend above because otherwise it covers most of the right-hand peak, but tumor samples are in green and normal tissue samples are in orange. No patterns particularly stand out to me here, except that sample 5 stands out even more when it's colored orange. I imagine that people who look at these plots more often might notice a pattern by sample type, but all I can see in these plots is the fact that sample 5 (R06C01) looks different from the rest.

### By sex

```{r}
sex = pData(rgSet)$Sex
densityPlot(rgSet,sampGroups = sex,legend = F)
densityBeanPlot(rgSet,sampGroups = sex)
```

Again, I had to remove the legend in the density plot because it covers up the right-hand peak, but in these plots females are in green and males in orange. Again, sample 5 stands out to me here but I don't see any patterns by sex. Sample 5 is not so different that it should be excluded, but I would try some between sample normalization and see if that helps it align more with the other samples.  

## d) Control probes

```{r}
controlStripPlot(rgSet,controls = "BISULFITE CONVERSION I",
                 sampNames = rgSet$id)
controlStripPlot(rgSet,controls = "NEGATIVE",
                 sampNames = rgSet$id)
```

Each 450K array includes internal control probes to help assess the quality of various pre-treatment steps. Bisulfite conversion probes are included to help assess the efficiency of bisulfite conversion, because "if the bisulfite conversion reaction was successful, the "C" (Converted) probes will match the converted sequence and get extended." Negative controls are included to help with background correction, by targeting bisulfite-converted sequences that don't contain CpGs. 

In the plots above, we want samples to be consistent within each of the red and green channels, so these plots look fairly reasonable. Samples 2 and 5 may look slightly different from the others, but it's difficult to say for sure based on just these plots. Also, the range of intensities is about the same for both the red and green channel, which is a good sign for this data because it indicates that there isn't significant red/green bias. Also, the intensity of the negative probes is generally lower than the bisulfite probes, which indicates that background noise should not be an issue with this dataset.

## e) Detection p values

```{r}
# Count p values >= 0.05 per sample
detect = detectionP(rgSet)
colSums(detect >= 0.05)
# Row means
rmeans = rowMeans(detect)
```

Detection p values indicate whether a probe's signal is greater than the average background signal. They are calculated as $p = 1-\phi[\frac{x-\mu_{neg}}{\sigma_{neg}}]$ where x is the sum of two beads (type I probes) or two color intensities (type II probes) and $\phi$ is the normal CDF.

Sample `r as.numeric(which.max(colSums(detect >= 0.05)))` has the most detection p values $\geq 0.05$ with `r max(colSums(detect >= 0.05))` (`r round(max(colSums(detect >= 0.05))/nrow(detect)*100,3)`%). Out of the `r nrow(detect)` probes, `r sum(rmeans >= 0.05)` have a mean p value $\geq 0.05$.

### Save the methylation signals

```{r}
mset = preprocessRaw(rgSet)
msetSWAN = preprocessSWAN(rgSet)
```

## f) Multidimensional scaling (MDS) plots

### By sex

#### SWAN-normalized

```{r}
mdsPlot(msetSWAN,sampGroups = sex,sampNames = rgSet$id,
        legendPos = "topleft")
mdsPlot(msetSWAN,numPositions = 10000,sampGroups = sex,
        sampNames = rgSet$id,legendPos = "topleft")
```

#### Raw data

```{r}
mdsPlot(mset,sampGroups = sex,sampNames = rgSet$id,legendPos = "topleft")
mdsPlot(mset,numPositions = 10000,sampGroups = sex,
        sampNames = rgSet$id,legendPos = "topright")
```

### By cancer status

#### SWAN-normalized

```{r}
mdsPlot(msetSWAN,sampGroups = stype,sampNames = rgSet$id,
        legendPos = "bottomleft")
mdsPlot(msetSWAN,numPositions = 10000,sampGroups = stype,
        sampNames = rgSet$id)
```

#### Raw data

```{r}
mdsPlot(mset,sampGroups = stype,sampNames = rgSet$id,
        legendPos = "topleft")
mdsPlot(mset,numPositions = 10000,sampGroups = stype,
        sampNames = rgSet$id,legendPos = "topright")
```

## g) Distribution of beta values before and after SWAN normalization

```{r fig.width=8,fig.height=10}
par(mfrow=c(6,2),mar=c(1,1,1,1))
for (i in 1:6) {
  plotBetasByType(mset[,i],main = paste(colnames(mset[,i]),"Before"))
  plotBetasByType(msetSWAN[,i],
                  main = paste(colnames(msetSWAN[,i]),"After"))
}
```

# 2. DNA Methylation Annotation and Differentially Methylated Positions (Illumina 450K)

Get genome annotation information:

```{r}
gset = mapToGenome(msetSWAN)
annotation = getAnnotation(gset)
```

## a) CpG islands, shores, shelves and open seas

```{r}
table(annotation$Relation_to_Island) %>% as.data.frame(.) %>%
  rename(.,Feature = Var1) %>% flextable(.)
```

## b) Find DMP for cancer status

```{r}
m = getM(msetSWAN)
pheno = pData(rgSet)$Status
dmp = dmpFinder(m,pheno = pheno,type = "categorical")
dmp %>% round(.,3) %>% rownames_to_column(var = "CpG") %>% 
  head(.,10) %>% flextable(.)
```

There are `r sum(dmp$qval <= 0.1)` DMPs with q-value $\leq 0.10$. In the DMP table above, a positive intercept indicates that the methylation was higher in the cancer samples compared to the normal tissue samples. Of the `r sum(dmp$pval <= 1e-5)` CpG sites significant at the p $\leq 10^{-5}$ level, `r sum(dmp$intercept[dmp$pval <= 1e-5] > 0)` was hypermethylated in cancer samples and `r sum(dmp$intercept[dmp$pval <= 1e-5] < 0)` were hypomethylated. This makes some sense, because I would expect that in general cancer cells would have higher gene expression than normal cells.

### Plot CpGs

```{r}
top4 = rownames(dmp[order(dmp$pval)[1:4],])
plotCpg(msetSWAN,cpg = top4,pheno)
```

## c) Find DMP for sex

```{r}
pheno = pData(rgSet)$Sex
dmp = dmpFinder(m,pheno = pheno,type = "categorical")
dmp %>% round(.,3) %>% rownames_to_column(var = "CpG") %>% 
  head(.,10) %>% flextable(.)
```

There are `r sum(dmp$qval <= 0.1)` DMPs with q-value $\leq 0.10$. Of the `r sum(dmp$pval <= 1e-5)` CpG sites significant at the p $\leq 10^{-5}$ level, `r sum(dmp$intercept[dmp$pval <= 1e-5] > 0)` was hypermethylated in females and `r sum(dmp$intercept[dmp$pval <= 1e-5] < 0)` were hypomethylated. 

### Plot CpGs

```{r}
top4 = rownames(dmp[order(dmp$pval)[1:4],])
plotCpg(msetSWAN,cpg = top4,pheno)
```

## d) Estimate whether each sample is male or female

```{r}
gset =mapToGenome(msetSWAN)
gset = addSex(gset)
pred_sex = 
  as.data.frame(cbind(pData(gset)$predictedSex, pData(gset)$Sex))
colnames(pred_sex) = c("Predicted Sex","Label")
flextable(pred_sex)
```

### Re-do MDS plot

```{r}
pred = pred_sex$`Predicted Sex`
mdsPlot(msetSWAN,sampGroups = pred)
mdsPlot(msetSWAN,numPositions = 10000,sampGroups = pred)
```

### Re-do 2c

```{r}
dmp = dmpFinder(m,pheno = pred,type = "categorical")
dmp %>% round(.,3) %>% rownames_to_column(var = "CpG") %>% 
  head(.,10) %>% flextable(.)
```

After correcting the sex labels, there are `r sum(dmp$qval <= 0.1)` DMPs with q-value $\leq 0.10$.

## e) bumphunter

```{r}
# Re-run for cancer status
pheno = pData(rgSet)$Status
dmp = dmpFinder(m,pheno = pheno,type = "categorical")
# Code from questions
diffs = dmp$intercept 
chr = annotation$chr
pos = annotation$pos
cl = clusterMaker(chr, pos, maxGap = 300) # cluster probes
#Find regions with a stretch of differences
segs = getSegments(diffs, f = cl, cutoff = 6)
#To plot the first region identified
j=1
ind = segs$dnIndex[[j]]
index = which(cl==cl[ind])
plot(pos[index],diffs[index],
xlab=paste("position on", chr[ind]), ylab="diff")
abline(h = 0, col = "blue")
```
